File: ios-uikit/DynamicTabBar/DynamicTabBar/NavigationComponent.swift
//
//  NavigationComponent.swift
//  DynamicTabBar
//
//  Bridge component for server-driven tab navigation.
//  Receives 'configure' messages from Rails and updates tab state.
//

import Foundation
import HotwireNative
import os.log

// MARK: - NavigationComponent

final class NavigationComponent: BridgeComponent {
    private let logger = Logger(subsystem: "com.example.dynamictabbar", category: "NavigationComponent")

    nonisolated override class var name: String { "navigation" }

    override func onReceive(message: Message) {
        guard message.event == "configure" else {
            logger.warning("Received unknown event: \(message.event)")
            return
        }

        // Decode the message data
        guard let data: MessageData = message.data() else {
            logger.error("Failed to decode message data")
            return
        }

        // Build directive from message data
        let directive: TabsDirective
        if data.tabs.isEmpty {
            directive = .bootstrap
        } else if let active = data.active {
            directive = .tabbed(active: active, tabs: data.tabs)
        } else {
            logger.error("Protocol violation: non-empty tabs without active field")
            return
        }

        // Post notification to TabbedNavigator
        NotificationCenter.default.post(
            name: .tabConfigurationDidChange,
            object: nil,
            userInfo: ["directive": directive]
        )

        let tabCount = directive.tabs.count
        let mode = tabCount == 0 ? "Bootstrap" : "Tabbed(\(tabCount))"
        logger.info("Sent tab config: \(mode)")
    }
}

// MARK: - Message Data

private extension NavigationComponent {
    struct MessageData: Codable {
        let active: String?
        let tabs: [TabData]
    }
}

File: android-fragments/app/src/main/java/io/tilde/dynamictabbar/NavigationComponent.kt
package io.tilde.dynamictabbar

import android.util.Log
import dev.hotwire.core.bridge.BridgeComponent
import dev.hotwire.core.bridge.BridgeDelegate
import dev.hotwire.core.bridge.Message
import dev.hotwire.navigation.destinations.HotwireDestination
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

private const val TAG = "Navigation"

/** Singleton for NavigationComponent → MainActivity communication. */
object TabDirectiveRouter {
    var listener: ((TabsDirective) -> Unit)? = null
    fun send(directive: TabsDirective) { listener?.invoke(directive) }
}

class NavigationComponent(
    name: String,
    private val delegate: BridgeDelegate<HotwireDestination>
) : BridgeComponent<HotwireDestination>(name, delegate) {

    override fun onReceive(message: Message) {
        when (message.event) {
            "configure" -> handleConfigure(message)
            else -> Log.w(TAG, "Received unknown event: ${message.event}")
        }
    }

    private fun handleConfigure(message: Message) {
        val data = try {
            json.decodeFromString<MessageData>(message.jsonData)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to decode message data")
            return
        }

        val directive = if (data.tabs.isEmpty()) {
            TabsDirective.Bootstrap
        } else {
            val active = data.active
            if (active == null) {
                Log.e(TAG, "Protocol violation: non-empty tabs without active field")
                return
            }
            TabsDirective.Tabbed(active = active, tabs = data.tabs)
        }

        TabDirectiveRouter.send(directive)

        val mode = if (directive.tabsList.isEmpty()) "Bootstrap" else "Tabbed(${directive.tabsList.size})"
        Log.d(TAG, "Sent tab config: $mode")
    }

    companion object {
        private val json = Json { ignoreUnknownKeys = true }
    }
}

@Serializable
private data class MessageData(
    val active: String? = null,
    val tabs: List<TabData> = emptyList()
)
